File: src/raycasting/stochasticRaytracing/sample4d.cpp
/**
4D vector sampling
*/

#include <cstdlib>

#include "common/error.h"
#include "common/quasiMonteCarlo/Halton.h"
#include "common/quasiMonteCarlo/ScrambledHalton.h"
#include "common/quasiMonteCarlo/Sobol.h"
#include "common/quasiMonteCarlo/Faure.h"
#include "common/quasiMonteCarlo/Niederreiter31.h"
#include "raycasting/stochasticRaytracing/sample4d.h"

static Sampler4DSequence seq = Sampler4DSequence::RANDOM;

static const char RANDOM_NAME[8] = "drand48";
static const char HALTON_NAME[7] = "Halton";
static const char SCRAMBLED_HALTON_NAME[12] = "ScramHalton";
static const char SOBOL_NAME[6] = "sobol";
static const char ORIGINAL_FAURE_NAME[6] = "faure";
static const char GENERALIZED_FAURE_NAME[7] = "GFaure";
static const char NIEDERREITER_NAME[5] = "Nied";
static const char UNKNOWN_NAME[8] = "Unknown";

static inline const char *
SEQ4D_NAME(Sampler4DSequence sequence) {
    switch ( sequence ) {
        case Sampler4DSequence::RANDOM:
            return RANDOM_NAME;
        case Sampler4DSequence::HALTON:
            return HALTON_NAME;
        case Sampler4DSequence::SCRAMBLED_HALTON:
            return SCRAMBLED_HALTON_NAME;
        case Sampler4DSequence::SOBOL:
            return SOBOL_NAME;
        case Sampler4DSequence::ORIGINAL_FAURE:
            return ORIGINAL_FAURE_NAME;
        case Sampler4DSequence::GENERALIZED_FAURE:
            return GENERALIZED_FAURE_NAME;
        case Sampler4DSequence::NIEDERREITER:
            return NIEDERREITER_NAME;
        default:
            return UNKNOWN_NAME;
    }
}

/**
Also initialises the sequence
*/
void
setSequence4D(Sampler4DSequence sequence) {
    seq = sequence;
    switch ( seq ) {
        case Sampler4DSequence::SOBOL:
            initSobol(4);
            break;
        case Sampler4DSequence::ORIGINAL_FAURE:
            initOriginalFaureSequence(4);
            break;
        case Sampler4DSequence::GENERALIZED_FAURE:
            initGeneralizedFaureSequence(4);
            break;
        default:
            break;
    }
}

/**
Returns 4D sample with given index from current sequence. When the
current sequence is 'random', the index is not used
*/
double *
sample4D(unsigned seed) {
    static double xi[4];
    const unsigned *zeta;
    const double *xx;

    switch ( seq ) {
        case Sampler4DSequence::RANDOM:
            xi[0] = drand48();
            xi[1] = drand48();
            xi[2] = drand48();
            xi[3] = drand48();
            break;
        case Sampler4DSequence::HALTON:
            xi[0] = Halton2((int)seed);
            xi[1] = Halton3((int)seed);
            xi[2] = Halton5((int)seed);
            xi[3] = Halton7((int)seed);
            break;
        case Sampler4DSequence::SCRAMBLED_HALTON:
            xx = scrambledHalton(seed, 4);
            xi[0] = xx[0];
            xi[1] = xx[1];
            xi[2] = xx[2];
            xi[3] = xx[3];
            break;
        case Sampler4DSequence::SOBOL:
            xx = sobol((int)seed);
            xi[0] = xx[0];
            xi[1] = xx[1];
            xi[2] = xx[2];
            xi[3] = xx[3];
            break;
        case Sampler4DSequence::ORIGINAL_FAURE:
        case Sampler4DSequence::GENERALIZED_FAURE:
            xx = faure((int) seed);
            xi[0] = xx[0];
            xi[1] = xx[1];
            xi[2] = xx[2];
            xi[3] = xx[3];
            break;
        case Sampler4DSequence::NIEDERREITER:
            zeta = niederreiter31(seed);
            xi[0] = (double) zeta[0] * RECIP;
            xi[1] = (double) zeta[1] * RECIP;
            xi[2] = (double) zeta[2] * RECIP;
            xi[3] = (double) zeta[3] * RECIP;
            break;
        default:
            logFatal(-1, "sample4D", "QMC Sequence %s not yet implemented", SEQ4D_NAME(seq));
    }

    return xi;
}

/**
The following routines are safe with Sample4D(), which calls only
31-bit sequences (including 31-bit Niederreiter sequence). If
you are looking for such a routine to use directly in conjunction
with the routine Nied() or NextNiedInRange(), you should use
the foldSample() routine in Niederreiter.h instead.
Nied() and NextNiedInRange() are 63-bit unless compiled without
'unsigned long long' support
*/
void
foldSampleU(unsigned *xi1, unsigned *xi2) {
    foldSample31(xi1, xi2);
}

void
foldSampleF(double *xi1, double *xi2) {
    unsigned zeta1 = (unsigned)(*xi1 * RECIP1);
    unsigned zeta2 = (unsigned)(*xi2 * RECIP1);
    foldSampleU(&zeta1, &zeta2);
    *xi1 = (double) zeta1 * RECIP;
    *xi2 = (double) zeta2 * RECIP;
}

File: java/src/main/java/vsdk/toolkit/raycasting/stochasticRaytracing/Sample4d.java
/**
4D vector sampling
*/

package vsdk.toolkit.raycasting.stochasticRaytracing;

import vsdk.toolkit.common.Error;
import vsdk.toolkit.numericalAnalysis.quasiMonteCarlo.Faure;
import vsdk.toolkit.numericalAnalysis.quasiMonteCarlo.Halton;
import vsdk.toolkit.numericalAnalysis.quasiMonteCarlo.Niederreiter31;
import vsdk.toolkit.numericalAnalysis.quasiMonteCarlo.ScrambledHalton;
import vsdk.toolkit.numericalAnalysis.quasiMonteCarlo.Sobol;

public class Sample4d {
    private static Sampler4DSequence seq = Sampler4DSequence.RANDOM;

    private static final String RANDOM_NAME = "drand48";
    private static final String HALTON_NAME = "Halton";
    private static final String SCRAMBLED_HALTON_NAME = "ScramHalton";
    private static final String SOBOL_NAME = "sobol";
    private static final String ORIGINAL_FAURE_NAME = "faure";
    private static final String GENERALIZED_FAURE_NAME = "GFaure";
    private static final String NIEDERREITER_NAME = "Nied";
    private static final String UNKNOWN_NAME = "Unknown";

    private static String sequenceName(Sampler4DSequence sequence) {
        switch ( sequence ) {
            case RANDOM:
                return RANDOM_NAME;
            case HALTON:
                return HALTON_NAME;
            case SCRAMBLED_HALTON:
                return SCRAMBLED_HALTON_NAME;
            case SOBOL:
                return SOBOL_NAME;
            case ORIGINAL_FAURE:
                return ORIGINAL_FAURE_NAME;
            case GENERALIZED_FAURE:
                return GENERALIZED_FAURE_NAME;
            case NIEDERREITER:
                return NIEDERREITER_NAME;
            default:
                return UNKNOWN_NAME;
        }
    }

    /**
Also initialises the sequence
*/
    public static void
    setSequence4D(Sampler4DSequence sequence) {
        seq = sequence;
        switch ( seq ) {
            case SOBOL:
                Sobol.initSobol(4);
                break;
            case ORIGINAL_FAURE:
                Faure.initOriginalFaureSequence(4);
                break;
            case GENERALIZED_FAURE:
                Faure.initGeneralizedFaureSequence(4);
                break;
            default:
                break;
        }
    }

    /**
Returns 4D sample with given index from current sequence. When the
current sequence is 'random', the index is not used
*/
    public static double[]
    sample4D(long seed) {
        double[] xi = new double[4];
        long[] zeta;
        double[] xx;

        switch ( seq ) {
            case RANDOM:
                xi[0] = Math.random();
                xi[1] = Math.random();
                xi[2] = Math.random();
                xi[3] = Math.random();
                break;
            case HALTON:
                xi[0] = Halton.Halton2((int)seed);
                xi[1] = Halton.Halton3((int)seed);
                xi[2] = Halton.Halton5((int)seed);
                xi[3] = Halton.Halton7((int)seed);
                break;
            case SCRAMBLED_HALTON:
                xx = ScrambledHalton.scrambledHalton(seed, 4);
                xi[0] = xx[0];
                xi[1] = xx[1];
                xi[2] = xx[2];
                xi[3] = xx[3];
                break;
            case SOBOL:
                xx = Sobol.sobol((int)seed);
                xi[0] = xx[0];
                xi[1] = xx[1];
                xi[2] = xx[2];
                xi[3] = xx[3];
                break;
            case ORIGINAL_FAURE:
            case GENERALIZED_FAURE:
                xx = Faure.faure((int)seed);
                xi[0] = xx[0];
                xi[1] = xx[1];
                xi[2] = xx[2];
                xi[3] = xx[3];
                break;
            case NIEDERREITER:
                zeta = Niederreiter31.niederreiter31(seed);
                xi[0] = zeta[0] * Niederreiter31.RECIP;
                xi[1] = zeta[1] * Niederreiter31.RECIP;
                xi[2] = zeta[2] * Niederreiter31.RECIP;
                xi[3] = zeta[3] * Niederreiter31.RECIP;
                break;
            default:
                Error.fatal(-1, "Sample4d::sample4D", "QMC Sequence %s not yet implemented", sequenceName(seq));
        }

        return xi;
    }

    /**
The following routines are safe with Sample4D(), which calls only
31-bit sequences (including 31-bit Niederreiter sequence). If
you are looking for such a routine to use directly in conjunction
with the routine Niederreiter::Nied() or Niederreiter::NextNiedInRange(), you should use
the Niederreiter::foldSample() routine in Niederreiter.h instead.
Niederreiter::Nied() and Niederreiter::NextNiedInRange() are 63-bit unless compiled without
'unsigned long long' support
*/
    public static void
    foldSampleU(long[] xi1, long[] xi2) {
        Niederreiter31.foldSample31(xi1, xi2);
    }

    public static void
    foldSampleF(double[] xi1, double[] xi2) {
        long zeta1 = (long)(xi1[0] * Niederreiter31.RECIP1);
        long zeta2 = (long)(xi2[0] * Niederreiter31.RECIP1);
        Sample4d.foldSampleU(new long[] {zeta1}, new long[] {zeta2});
        xi1[0] = zeta1 * Niederreiter31.RECIP;
        xi2[0] = zeta2 * Niederreiter31.RECIP;
    }
}
